### ISSUE TITLE
Base class is should not be detected as a static holder type

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn-analyzers#1820]
Analyzer package: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)
Package Version: 2.6.2-beta2
Diagnostic ID: CA1052

This is something I do fairly commonly to organize tests:

```c#
[Category("Integration")]
public abstract class IntegrationTests
{
    protected static SqlConnection CreateConnection() => // ...

   // ...
}

// Has category Integration due to inheritance
public sealed class ReportingTests : IntegrationTests
{
    public static void SomeTest()
    {
        // Name CreateConnection is in scope due to inheritance
    }
}
```

> CA1052 Type 'IntegrationTests' is a static holder type but is neither static nor NotInheritable

Since `IntegrationTests` is abstract, or at minimum since `ReportingTests` is deriving from it, it should not be considered a static holder type.

**Comment by sharwell** [comment]
I'm not sure I agree with the original issue. My concern is the analyzer would no longer report diagnostics in cases where a user put the `abstract` modifier on a type where the `static` modifier is a better choice. I believe this situation is a primary target of the analyzer.

**Comment by jnm2** [comment]
@sharwell Why not have the analyzer bail when it sees that `ReportingTests` inherits from it? The modifier it's suggesting would actually break the compilation.

**Comment by sharwell** [comment]
> Why not have the analyzer bail when it sees that `ReportingTests` inherits from it?

This forces the analyzer to be a Compilation End analyzer, which is disabled in most scenarios and crippled in the remaining ones. Symbol end analyzers provide better overall behavior but are restricted to the contents of a single type.

**Comment by mahalex** [comment]
My reasoning for making the change derives from Microsoft's [documentation on CA1052](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1052-static-holder-types-should-be-sealed), which reads

> Suppress a warning from this rule only if the type is designed to be inherited.

To me, if a class is marked as `abstract`, it is really designed to be inherited, so the warning should not appear.

I can kind of understand the argument of @sharwell about the situation where a user puts `abstract` instead of `static` on a class, but I don't think it happens often enough in real-life scenarios: these keywords are quite different in semantics.

**Comment by sharwell** [comment]
@mahalex My interpretation of that line is "In this situation, a warning will still be reported, but the developer may suppress it rather than change the code to avoid the warning." Changing the diagnostic is equivalent to *forcing* the suppression in this case.

**Comment by mavasani** [comment]
Overall, I agree with @sharwell's stance here, which seems to match the documentation as it never explicitly calls out that abstract types are not flagged. However, legacy FxCop rule implementation for CA1052 explicitly bails out for abstract types, so we have to do the same for compat reason. Once we implement #1850 (or allow third party analyzer options in .editorconfig), we can likely add a knob to allow end users to configure their preference.
